components/ledger/internal/bootstrap/balance_sync_collector.go

```
@@ -0,0 +1,270 @@
+
+	c.flushFn = fn
+}
+
+// Run starts the collector's main loop. It polls for eligible keys and flushes
+// based on the dual-trigger (size OR timeout). Blocks until context is cancelled.
+//
+// The loop has three modes:
+//   - Busy mode: keys are available, tight poll loop to accumulate quickly
+//   - Draining mode: buffer has items but no new keys, wait for timeout or poll
+//   - Idle mode: nothing to do, sleep until next scheduled key
+func (c *BalanceSyncCollector) Run(ctx context.Context, fetchKeys FetchKeysFunc, waitForNext WaitForNextFunc) {
```

golangci-lint (2.11.4)

[error] cognitive complexity 45 of func `(*BalanceSyncCollector).Run` is high (> 30)

(gocognit)